Documentation/MyFirstContribution

```diff
@@ -0,0 +1,584 @@
+suggests the best place to clone from is GitHub.
+
+----
+git clone https://github.com/git/git git
+----
+
+=== Identify Problem to Solve
+
+In this codelab, we will add a new command, `git psuh`, short for "Pony Saying
+`Um, Hello'" - a feature which has gone unimplemented despite a high frequency
+of invocation during users' typical daily workflow.
+
+(We've seen some other effort in this space with the implementation of popular
+commands such as `sl`.)
+
+=== Set Up Your Workspace
+
+Let's start by making a development branch to work on our changes. Per
+`Documentation/SubmittingPatches`, since a brand new command is a new feature,
+it's fine to base your work on `master`. However, in the future for bugfixes,
+etc., you should check that doc and base it on the appropriate branch.
+
+For the purposes of this doc, we will base all our work on `master`. Before
+running the command below, ensure that you are on `master` first so your
+branch diverges at the right point.
+
+----
+git checkout -b psuh
+----
+
+We'll make a number of commits here in order to demonstrate how to send many
+patches up for review simultaneously.
+
+== Code It Up!
+
+NOTE: A reference implementation can be found at
+https://github.com/nasamuffin/git/tree/codelab.
+
+=== Adding a new command
+
+Lots of the main useful commands are written as builtins. So it is informative
+to implement `git psuh` as a builtin.
+
+Create a new file in `builtin/` called `psuh.c`.
+
+The entry point of your new command needs to match a certain signature:
+
+----
+int cmd_psuh(int argc, const char **argv, const char *prefix)
+----
+
+We'll also need to add the extern declaration of psuh; open up `builtin.h`,
+find the declaration for cmd_push, and add a new line for psuh:
+
+----
+extern int cmd_psuh(int argc, const char **argv, const char *prefix);
+----
+
+Be sure to `#include "builtin.h"` in your `psuh.c`.
+
+Go ahead and add some throwaway printf to that method. This is a decent
+starting point as we can now add build rules and register the command.
+
+Let's try to build it.  Open Makefile, find where `builtin/push.o` is added
+to BUILTIN_OBJS, and add `builtin/psuh.o` in the same way. Once you've done so,
+move to the root directory and build simply with `make -j16`. Optionally, add
+the DEVELOPER=1 variable to turn on some additional warnings:
+
+----
+echo DEVELOPER=1 > config.mak
+make -j16
+----
+
+Great, now your new command builds happily on its own. But nobody invokes it.
+Let's change that.
+
+The list of commands lives in `git.c`. We can register a new command by adding
+a cmd_struct to the commands[] array. struct cmd_struct takes a string with the
+command name, a function pointer to the command implementation, and a setup
+option flag. For now, let's keep cheating off of push. Find the line where
+cmd_push is registered, copy it, and modify it for cmd_psuh. Leave the options
+the same as for push.
+
+TODO: what do the options do? How to fix them?
```

Comment:
The flags mean the following:

<dl>
<dt>RUN_SETUP</dt>
<dd>The command requires a Git repository (which can be bare, or more commonly be associated with a worktree).</dd>
<dt>RUN_SETUP_GENTLY</dt>
<dd>The command does not require a Git repository, but uses it if available. This flag should be used also for commands that conditionally require a Git directory, in which case they can test for it via `startup_info.have_repository`.</dd>
<dt>USE_PAGER</dt>
<dd>The command produces potentially a lot of output which needs to be paged (like `git log`).</dd>
<dt>NEED_WORK_TREE</dt>
<dd>The command requires a worktree, i.e. cannot work on a bare repository (e.g. `git merge`). Requires RUN_SETUP</dd>
<dt>SUPPORT_SUPER_PREFIX</dt>
<dd>The command can prefix paths in submodules with the prefix as per the super project (specified via the environment variable `GIT_INTERNAL_SUPER_PREFIX`).</dd>
<dt>DELAY_PAGER_CONFIG</dt>
<dd>The command wants to set up the pager manually. This comes in handy for commands which have some subcommands that require the output to paged, and other subcommands that do not.</dd>
<dt>NO_PARSEOPT</dt>
<dd>The command does *not* use the `parse-options` machinery to parse its command-line options. New commands *should* use `parse-options`.</dd>
</dl>